### DISCUSSION TITLE
Dynamic validation based on user-generated data

### DISCUSSION THREAD

**Discussion by austinzentzkippdc** [colinhacks/zod#2834] [Q&A] [answered]
Hi!

I'm considering a swap from Yup to Zod, as I appreciate the syntax and the community. But I have a few validation scenarios in Yup that involve dynamic validation, and I'm curious how I'd accomplish similar goals in Zod.

Here's a scenario:

> Imagine a database of students and colleges. Admins can create college profiles, and then can add students to each college. However, the required data for each student varies depending on the college they attend. for example:
> * Some colleges want to record the student's state of residence
> * Some colleges want to note the student's high school GPA
> * Some colleges want to note legacy status.
>
> An admin might add two colleges: State U, which records the state of residence and the GPA but _not_ legacy status. And Ivy U, which disregards the state of residence but wants to note the GPA and legacy status.

So, I might have a generic Zod schema like this:

```ts
const input = z.object({
  firstName: z.string(),
  lastName: z.string(),
  schoolId: z.string(),
  state: z.string().optional(),
  gpa: z.number().optional(),
  isLegacy: z.boolean().optional()
})
```

But, depending on the school chosen, `state`, `gpa`, and `isLegacy` may all be required fields.

With yup, I can pass the school object in as context, and then modify which fields are required accordingly. (This is accomplished by using `'$context'` strings.)

I'm not quite sure what my options would be in Yup. Would unions/discriminated unions be suitable for this use case? Those look similar but I haven't been able to determine whether they serve quite the same function.

Or would I need to architect my data differently? Perhaps do the structure of the validation in Zod, and then turn to some other library or custom function to manage the conditional validation?

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const shared = z.object( {
    firstName: z.string(),
    lastName: z.string(),
} )

const schema = z.discriminatedUnion( 'schoolId', [
    shared.extend( {
        schoolId: z.literal( 'State U' ),
        state: z.string(),
        gpa: z.number(),
    } ),
    shared.extend( {
        schoolId: z.literal( 'Ivy U' ),
        gpa: z.number(),
        isLegacy: z.boolean(),
    } ),
] )

console.log(
    schema.safeParse( {
        firstName: 'John',
        lastName: 'Doe',
        schoolId: 'State U',
        state: 'CA',
        gpa: 3.5,
    } ).success
) // true

console.log(
    schema.safeParse( {
        firstName: 'John',
        lastName: 'Doe',
        schoolId: 'Ivy U',
        gpa: 3.5,
        isLegacy: true,
    } ).success
) // true
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
